File: dashboard/scripts/generate-reference.mjs
/**
 * Generate docs/DASHBOARD-REFERENCE.md from dashboard codebase.
 *
 * Extracts file-level JSDoc comments (module docstrings), test descriptions
 * (describe/it strings), and intentional/tech-debt annotations from .ts/.tsx
 * files. Mirrors the backend pattern in generate_reference.py.
 *
 * Usage:
 *     node dashboard/scripts/generate-reference.mjs
 */

import { readFileSync, writeFileSync, readdirSync, statSync, mkdirSync } from "node:fs"
import { join, relative, extname } from "node:path"

const SCRIPT_DIR = new URL(".", import.meta.url).pathname
const DASHBOARD_DIR = join(SCRIPT_DIR, "..")
const REPO_ROOT = join(DASHBOARD_DIR, "..")
const OUTPUT_PATH = join(REPO_ROOT, "docs", "DASHBOARD-REFERENCE.md")

const SKIP_DIRS = new Set(["node_modules", ".next", "__pycache__", ".turbo", "dist"])
const TS_EXTS = new Set([".ts", ".tsx"])

// ── Helpers ──────────────────────────────────────────────────────────────

function walkFiles(dir, exts) {
  const results = []
  let entries
  try {
    entries = readdirSync(dir)
  } catch {
    return results
  }
  for (const entry of entries.sort()) {
    const full = join(dir, entry)
    let st
    try {
      st = statSync(full)
    } catch {
      continue
    }
    if (st.isDirectory()) {
      if (!SKIP_DIRS.has(entry)) {
        results.push(...walkFiles(full, exts))
      }
    } else if (exts.has(extname(entry))) {
      results.push(full)
    }
  }
  return results
}

function relPath(absPath) {
  return relative(REPO_ROOT, absPath)
}

// ── Extractors ───────────────────────────────────────────────────────────

/**
 * Extract the first JSDoc block if it starts at the very top of the file
 * (possibly preceded by blank lines or single-line comments like "use client").
 * Returns { content, line } or null.
 */
function extractFileJSDoc(src) {
  // Find the first /** ... */ block. It must appear before any import/export
  // or code statement to count as a "module docstring".
  const match = src.match(/^((?:\s*(?:\/\/[^\n]*)?\n)*)\s*(\/\*\*[\s\S]*?\*\/)/)
  if (!match) return null

  const preamble = match[1]
  const block = match[2]

  // Ensure nothing code-like precedes it (only whitespace, single-line comments,
  // or "use client"/"use server" directives are allowed before the JSDoc).
  const preambleLines = preamble.split("\n")
  for (const line of preambleLines) {
    const trimmed = line.trim()
    if (
      trimmed === "" ||
      trimmed.startsWith("//") ||
      /^["']use (client|server)["'];?$/.test(trimmed)
    ) {
      continue
    }
    // Something else precedes the JSDoc — not a file-level docstring
    return null
  }

  // Strip the /** and */ delimiters and leading " * " from each line
  const inner = block
    .replace(/^\/\*\*\s*/, "")
    .replace(/\s*\*\/$/, "")
    .split("\n")
    .map((l) => l.replace(/^\s*\*\s?/, ""))
    .join("\n")
    .trim()

  if (!inner) return null

  // Line number of the opening /**
  const line = preamble.split("\n").length
  return { content: inner, line }
}

/**
 * Extract // intentional: and // tech-debt: annotations.
 * Best-effort skip of comments inside string literals.
 */
function extractAnnotations(src, filePath) {
  const chunks = []
  const lines = src.split("\n")
  const pattern = /\/\/\s*(intentional|tech-debt):\s*(.+)/

  // Build a set of line numbers that are inside template literals or
  // multi-line strings. This is best-effort — no full parser.
  const inString = new Set()
  let inTemplateLiteral = false
  let braceDepth = 0
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    // Very rough: track backtick-delimited regions
    const backticks = (line.match(/`/g) || []).length
    if (inTemplateLiteral) {
      inString.add(i + 1)
      if (backticks % 2 === 1) inTemplateLiteral = false
    } else if (backticks % 2 === 1) {
      // Opens a template literal that doesn't close on this line
      // But if it also closes, it's a single-line template — don't mark
      inTemplateLiteral = true
    }
  }

  for (let i = 0; i < lines.length; i++) {
    if (inString.has(i + 1)) continue
    const match = pattern.exec(lines[i])
    if (!match) continue

    const tag = match[1]
    const reason = match[2].trim()
    const rel = relPath(filePath)
    const lineNo = i + 1
    chunks.push({
      fileName: filePath.split("/").pop(),
      lineNo,
      content: reason,
      source: `${rel}:${lineNo}`,
      type: tag === "intentional" ? "annotation" : "tech_debt",
    })
  }
  return chunks
}

/**
 * Extract describe("...", ...) and it("...", ...) strings from test files.
 * Returns nested structure: [{ describe, its: [...] }]
 */
function extractTestDescriptions(src, filePath) {
  const chunks = []
  const rel = relPath(filePath)

  // Match describe("...", and it("...", at statement level (leading whitespace only).
  // The (?:^|\n) anchor + optional whitespace prevents matching .split("\n") etc.
  const describeRe = /(?:^|\n)\s*describe\(\s*["'`]((?:[^"'`\\]|\\.)*)["'`]/g
  const itRe = /(?:^|\n)\s*it\(\s*["'`]((?:[^"'`\\]|\\.)*)["'`]/g

  let match
  while ((match = describeRe.exec(src)) !== null) {
    const lineNo = src.slice(0, match.index).split("\n").length
    chunks.push({
      text: match[1],
      lineNo,
      source: `${rel}:${lineNo}`,
      kind: "describe",
    })
  }
  while ((match = itRe.exec(src)) !== null) {
    const lineNo = src.slice(0, match.index).split("\n").length
    chunks.push({
      text: match[1],
      lineNo,
      source: `${rel}:${lineNo}`,
      kind: "it",
    })
  }
  return chunks
}

// ── Main ─────────────────────────────────────────────────────────────────

function main() {
  const allFiles = walkFiles(DASHBOARD_DIR, TS_EXTS)
  const modules = []
  const annotations = []
  const techDebt = []
  const testPrinciples = [] // grouped by file

  for (const filePath of allFiles) {
    const rel = relPath(filePath)
    let src
    try {
      src = readFileSync(filePath, "utf-8")
    } catch {
      continue
    }

    // Skip test files for module docstrings and annotations
    const isTest =
      rel.includes("__tests__") ||
      filePath.endsWith(".test.ts") ||
      filePath.endsWith(".test.tsx")

    // Module docstrings (non-test files only)
    if (!isTest) {
      const doc = extractFileJSDoc(src)
      if (doc) {
        modules.push({
          title: rel,
          content: doc.content,
          source: `${rel}:${doc.line}`,
        })
      }
    }

    // Annotations (non-test files only)
    if (!isTest) {
      const annots = extractAnnotations(src, filePath)
      for (const a of annots) {
        if (a.type === "annotation") annotations.push(a)
        else techDebt.push(a)
      }
    }

    // Test descriptions (test files only)
    if (isTest) {
      const descs = extractTestDescriptions(src, filePath)
      if (descs.length > 0) {
        // Also extract file-level JSDoc as test file summary
        const doc = extractFileJSDoc(src)
        testPrinciples.push({
          file: rel,
          summary: doc ? doc.content : null,
          descriptions: descs,
        })
      }
    }
  }

  // ── Render ───────────────────────────────────────────────────────────

  const lines = [
    "# Agentobox Dashboard Reference",
    "",
    "> Auto-generated from codebase. Do not edit — regenerate with `node dashboard/scripts/generate-reference.mjs`.",
    "",
  ]

  // Modules
  if (modules.length > 0) {
    lines.push("## Modules")
    lines.push("")
    for (const mod of modules) {
      lines.push(`### ${mod.title}`)
      lines.push("")
      lines.push(mod.content)
      lines.push("")
    }
  }

  // Test Principles
  if (testPrinciples.length > 0) {
    lines.push("## Test Principles")
    lines.push("")
    for (const tp of testPrinciples) {
      lines.push(`### ${tp.file}`)
      lines.push("")
      if (tp.summary) {
        lines.push(tp.summary)
        lines.push("")
      }
      // Group: describe blocks with their it() children
      const describes = tp.descriptions.filter((d) => d.kind === "describe")
      const its = tp.descriptions.filter((d) => d.kind === "it")

      if (describes.length > 0) {
        for (const desc of describes) {
          lines.push(`**${desc.text}**`)
          // Find it() blocks that appear after this describe and before the next
          const nextDescLine =
            describes.find((d) => d.lineNo > desc.lineNo)?.lineNo ?? Infinity
          const childIts = its.filter(
            (i) => i.lineNo > desc.lineNo && i.lineNo < nextDescLine
          )
          if (childIts.length > 0) {
            for (const it of childIts) {
              lines.push(`- ${it.text}`)
            }
          }
          lines.push("")
        }
      } else if (its.length > 0) {
        // Standalone it() blocks without describe wrapper
        for (const it of its) {
          lines.push(`- ${it.text}`)
        }
        lines.push("")
      }
    }
  }

  // Exception Annotations
  if (annotations.length > 0) {
    lines.push("## Exception Annotations")
    lines.push("")
    lines.push("| File | Line | Annotation |")
    lines.push("|------|------|------------|")
    for (const a of annotations) {
      lines.push(`| ${a.fileName} | ${a.lineNo} | ${a.content} |`)
    }
    lines.push("")
  }

  // Tech Debt
  if (techDebt.length > 0) {
    lines.push("## Tech Debt")
    lines.push("")
    lines.push("| File | Line | Annotation |")
    lines.push("|------|------|------------|")
    for (const a of techDebt) {
      lines.push(`| ${a.fileName} | ${a.lineNo} | ${a.content} |`)
    }
    lines.push("")
  }

  const content = lines.join("\n")

  mkdirSync(join(REPO_ROOT, "docs"), { recursive: true })
  writeFileSync(OUTPUT_PATH, content, "utf-8")

  const totalChunks =
    modules.length +
    testPrinciples.length +
    annotations.length +
    techDebt.length
  process.stderr.write(
    `Wrote ${OUTPUT_PATH} (${totalChunks} chunks: ${modules.length} modules, ${testPrinciples.length} test files, ${annotations.length} annotations, ${techDebt.length} tech-debt)\n`
  )
}

main()

File: agent/scripts/generate-reference.py
"""Generate docs/AGENT-REFERENCE.md from agent codebase.

Walks agent/{runtime,transports,contracts,platform}/**/*.py, extracts module
docstrings, test class docstrings, and # intentional: / # tech-debt: annotations.
Mirrors the backend pattern in generate_reference.py.

Usage:
    python agent/scripts/generate-reference.py
"""

import ast
import re
import sys
from dataclasses import dataclass
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parent
AGENT_DIR = SCRIPT_DIR.parent
REPO_ROOT = AGENT_DIR.parent
OUTPUT_PATH = REPO_ROOT / "docs" / "AGENT-REFERENCE.md"

SCAN_DIRS = ["runtime", "transports", "contracts", "platform"]
SKIP_DIRS = {"__pycache__", ".venv", "node_modules"}


@dataclass
class DocChunk:
    title: str
    content: str
    source: str
    chunk_type: str  # "module" | "test_principle" | "annotation" | "tech_debt"


def _rel(path: Path) -> str:
    try:
        return str(path.relative_to(REPO_ROOT))
    except ValueError:
        return str(path)


def _should_walk(path: Path) -> bool:
    return not any(part in SKIP_DIRS for part in path.parts)


def _extract_module_docstrings() -> list[DocChunk]:
    chunks = []
    for subdir in SCAN_DIRS:
        scan_path = AGENT_DIR / subdir
        if not scan_path.is_dir():
            continue
        for py_file in sorted(scan_path.rglob("*.py")):
            if not _should_walk(py_file):
                continue
            if py_file.name == "__init__.py":
                src = py_file.read_text(encoding="utf-8").strip()
                if not src:
                    continue
            try:
                tree = ast.parse(py_file.read_text(encoding="utf-8"))
            except SyntaxError:
                continue
            docstring = ast.get_docstring(tree)
            if not docstring:
                continue
            rel = _rel(py_file)
            chunks.append(DocChunk(
                title=rel, content=docstring.strip(),
                source=f"{rel}:1", chunk_type="module",
            ))
    return chunks


def _extract_test_principles() -> list[DocChunk]:
    chunks = []
    tests_dir = AGENT_DIR / "tests"
    if not tests_dir.is_dir():
        return chunks
    for py_file in sorted(tests_dir.glob("test_*.py")):
        try:
            tree = ast.parse(py_file.read_text(encoding="utf-8"))
        except SyntaxError:
            continue
        for node in ast.walk(tree):
            if not isinstance(node, ast.ClassDef):
                continue
            docstring = ast.get_docstring(node)
            if not docstring:
                continue
            rel_file = py_file.name
            chunks.append(DocChunk(
                title=f"{rel_file} — {node.name}",
                content=docstring.strip(),
                source=f"agent/tests/{rel_file}:{node.lineno}",
                chunk_type="test_principle",
            ))
    return chunks


def _extract_annotations() -> list[DocChunk]:
    pattern = re.compile(r"#\s*(intentional|tech-debt):\s*(.+)")
    chunks = []
    for subdir in SCAN_DIRS:
        scan_path = AGENT_DIR / subdir
        if not scan_path.is_dir():
            continue
        for py_file in sorted(scan_path.rglob("*.py")):
            if not _should_walk(py_file) or "tests" in py_file.parts:
                continue
            src = py_file.read_text(encoding="utf-8")
            lines = src.splitlines()
            rel = _rel(py_file)

            string_lines: set[int] = set()
            try:
                tree = ast.parse(src)
                for node in ast.walk(tree):
                    if isinstance(node, ast.Constant) and isinstance(node.value, str):
                        if hasattr(node, "lineno") and hasattr(node, "end_lineno"):
                            for ln in range(node.lineno, (node.end_lineno or node.lineno) + 1):
                                string_lines.add(ln)
            except SyntaxError:
                continue

            for i, line in enumerate(lines, 1):
                if i in string_lines:
                    continue
                match = pattern.search(line)
                if match:
                    tag = match.group(1)
                    chunk_type = "annotation" if tag == "intentional" else "tech_debt"
                    chunks.append(DocChunk(
                        title=f"{py_file.name}:{i}",
                        content=match.group(2).strip(),
                        source=f"{rel}:{i}",
                        chunk_type=chunk_type,
                    ))
    return chunks


def _render_markdown(chunks: list[DocChunk]) -> str:
    lines = [
        "# Agentobox Agent Reference",
        "",
        "> Auto-generated from agent codebase. Do not edit — regenerate with `make docs`.",
        "",
    ]

    modules = [c for c in chunks if c.chunk_type == "module"]
    if modules:
        lines.append("## Modules")
        lines.append("")
        for chunk in modules:
            lines.append(f"### {chunk.title}")
            lines.append("")
            lines.append(chunk.content)
            lines.append("")

    principles = [c for c in chunks if c.chunk_type == "test_principle"]
    if principles:
        lines.append("## Test Principles")
        lines.append("")
        for chunk in principles:
            lines.append(f"### {chunk.title}")
            lines.append("")
            lines.append(chunk.content)
            lines.append("")

    annotations = [c for c in chunks if c.chunk_type == "annotation"]
    if annotations:
        lines.append("## Exception Annotations")
        lines.append("")
        lines.append("| File | Line | Annotation |")
        lines.append("|------|------|------------|")
        for chunk in annotations:
            parts = chunk.title.split(":")
            fname = parts[0]
            lineno = parts[1] if len(parts) > 1 else "?"
            lines.append(f"| {fname} | {lineno} | {chunk.content} |")
        lines.append("")

    tech_debt = [c for c in chunks if c.chunk_type == "tech_debt"]
    if tech_debt:
        lines.append("## Tech Debt")
        lines.append("")
        lines.append("| File | Line | Annotation |")
        lines.append("|------|------|------------|")
        for chunk in tech_debt:
            parts = chunk.title.split(":")
            fname = parts[0]
            lineno = parts[1] if len(parts) > 1 else "?"
            lines.append(f"| {fname} | {lineno} | {chunk.content} |")
        lines.append("")

    return "\n".join(lines)


def main():
    chunks = []
    chunks.extend(_extract_module_docstrings())
    chunks.extend(_extract_test_principles())
    chunks.extend(_extract_annotations())
    content = _render_markdown(chunks)

    OUTPUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    OUTPUT_PATH.write_text(content, encoding="utf-8")

    counts = {}
    for c in chunks:
        counts[c.chunk_type] = counts.get(c.chunk_type, 0) + 1
    parts = [f"{v} {k}s" for k, v in sorted(counts.items())]
    print(f"Wrote {OUTPUT_PATH} ({len(chunks)} chunks: {', '.join(parts)})")


if __name__ == "__main__":
    main()
